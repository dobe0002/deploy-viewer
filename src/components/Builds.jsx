import React from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import { Container } from 'reactstrap';
import BuildRow from './BuildRow';
import { getStartEndDate, convertDateToString } from '../js/getDates';

import FilterDateAppGroup from './FilterDateAppGroup';

// TODO:  Need to fully incorporate the FilterDateAppGroup here

export default class Builds extends React.Component {
  constructor(props) {
    super(props);
    const { currentAppGroup, deployData, startDate, endDate } = this.props;
    const dateRange = getStartEndDate(deployData);

    const sDate = startDate || dateRange.min;
    const eDate = endDate || dateRange.max;
    this.state = {
      startDate: sDate,
      endDate: eDate,
      currentAppGroup
    };
    this.handleAppGroupChange = this.handleAppGroupChange.bind(this);
    this.handleStartChange = this.handleStartChange.bind(this);
    this.handleEndChange = this.handleEndChange.bind(this);
    this.setHistory = this.setHistory.bind(this);
  }

  setHistory() {
    const { history } = this.props;
    const { startDate, endDate, currentAppGroup } = this.state;
    const sDate = convertDateToString(startDate);
    const eDate = convertDateToString(endDate);

    history.push(
      `/deploys/${currentAppGroup}/startDate/${sDate}/endDate/${eDate}`
    );
  }

  handleAppGroupChange(newAppGroup) {
    this.setState({ currentAppGroup: newAppGroup }, () => {
      this.setHistory();
    });
  }

  handleStartChange(date) {
    this.setState({ startDate: date }, () => {
      this.setHistory();
    });
  }

  handleEndChange(date) {
    this.setState({ endDate: date }, () => {
      this.setHistory();
    });
  }

  render() {
    const { appGroups, deployData } = this.props;
    const { currentAppGroup, startDate, endDate } = this.state;
    return (
      <Container fluid>
        <FilterDateAppGroup
          dataTitle='Deploy Status'
          startDate={startDate}
          endDate={endDate}
          currentAppGroup={currentAppGroup}
          appGroups={appGroups}
          onStartDateChange={this.handleStartChange}
          onEndDateChange={this.handleEndChange}
          onAppGroupChange={this.handleAppGroupChange}
        />

        {map(deployData, app => {
          if (currentAppGroup === '' || currentAppGroup === app.appGroup) {
            return (
              <BuildRow
                appTitle={app.title}
                appStatus={app.status}
                startDate={startDate}
                endDate={endDate}
              />
            );
          }
          return '';
        })}
      </Container>
    );
  }
}

Builds.propTypes = {
  deployData: PropTypes.arrayOf,
  appGroups: PropTypes.objectOf,
  currentAppGroup: PropTypes.string,
  history: PropTypes.objectOf,
  startDate: PropTypes.instanceOf(Date),
  endDate: PropTypes.instanceOf(Date)
};
Builds.defaultProps = {
  deployData: [],
  appGroups: {},
  currentAppGroup: '',
  history: { push: () => {} },
  startDate: undefined,
  endDate: undefined
};
