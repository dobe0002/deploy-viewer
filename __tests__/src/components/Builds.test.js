/* eslint-disable react/jsx-filename-extension */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Builds from '../../../src/components/Builds';
import deployData from '../../../__fixtures__/getDataFormat/sheetData';

configure({ adapter: new Adapter() });

const { axe, toHaveNoViolations } = require('jest-axe');

expect.extend(toHaveNoViolations);

describe('Builds  tests', () => {
  const sampleProps = {
    deployData
  };
  test('Accessibility check', async () => {
    const wrapper = shallow(<Builds deployData={sampleProps.deployData} />);
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  test('Change in appgroup changes number of builds ', () => {
    const wrapper = shallow(<Builds deployData={sampleProps.deployData} />);
    wrapper.instance().handleAppGroupChange('HR');
    expect(wrapper.find('BuildRow')).toHaveLength(2);

    wrapper.instance().handleAppGroupChange('');
    expect(wrapper.find('BuildRow')).toHaveLength(4);
  });
  test('Call HandleStartChange', () => {
    const wrapper = shallow(<Builds deployData={sampleProps.deployData} />);
    expect(wrapper.state('startDate')).toEqual(
      new Date('2019-11-01T06:00:00.000Z')
    );

    const newDate = new Date(2019, 11, 12);
    wrapper.instance().handleStartChange(newDate);
    expect(wrapper.state('startDate')).toEqual(newDate);
  });

  test('Call HandleEndChange', () => {
    const wrapper = shallow(<Builds deployData={sampleProps.deployData} />);
    expect(wrapper.state('endDate')).toEqual(
      new Date('2019-12-15T06:00:00.000Z')
    );

    const newDate = new Date('2019-12-12T06:00:00.000Z');
    wrapper.instance().handleEndChange(newDate);
    expect(wrapper.state('endDate')).toEqual(newDate);
  });

  test('Snapshot test', () => {
    const wrapper = shallow(<Builds deployData={sampleProps.deployData} />);
    // check appGroup and # apps is correct
    expect(wrapper.find('BuildRow')).toHaveLength(4);
    expect(wrapper.state('currentAppGroup')).toEqual('');

    expect(wrapper).toMatchSnapshot();
  });
});
