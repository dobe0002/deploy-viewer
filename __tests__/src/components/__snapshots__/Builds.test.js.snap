// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builds  tests Snapshot test 1`] = `
<Container
  fluid={true}
  tag="div"
>
  <FilterDateAppGroup
    appGroups={Object {}}
    currentAppGroup=""
    dataTitle="Deploy Status"
    endDate={2019-12-15T06:00:00.000Z}
    onAppGroupChange={[Function]}
    onEndDateChange={[Function]}
    onStartDateChange={[Function]}
    startDate={2019-11-01T05:00:00.000Z}
  />
  <BuildRow
    appStatus={
      Array [
        Object {
          "date": 2019-12-01T06:00:00.000Z,
          "dateSort": 20191201,
          "dateString": "12/01/2019",
          "note": "Users could not log in",
          "row": 10,
          "status": "C",
          "statusDescription": "Deployment with Customer Impact",
        },
        Object {
          "date": 2019-12-01T06:00:00.000Z,
          "dateSort": 20191201,
          "dateString": "12/01/2019",
          "note": "Fixed issues from the morning deployment",
          "row": 11,
          "status": "G",
          "statusDescription": "Good deployment",
        },
        Object {
          "date": 2019-12-15T06:00:00.000Z,
          "dateSort": 20191215,
          "dateString": "12/15/2019",
          "note": ":-)",
          "row": 9,
          "status": "G",
          "statusDescription": "Good deployment",
        },
      ]
    }
    appTitle="Cap"
    endDate={2019-12-15T06:00:00.000Z}
    startDate={2019-11-01T05:00:00.000Z}
  />
  <BuildRow
    appStatus={
      Array [
        Object {
          "date": 2019-11-15T06:00:00.000Z,
          "dateSort": 20191115,
          "dateString": "11/15/2019",
          "note": "Let's try some \\"BOLD\\" text",
          "row": 8,
          "status": "C",
          "statusDescription": "Deployment with Customer Impact",
        },
      ]
    }
    appTitle="Ethos"
    endDate={2019-12-15T06:00:00.000Z}
    startDate={2019-11-01T05:00:00.000Z}
  />
  <BuildRow
    appStatus={
      Array [
        Object {
          "date": 2019-11-02T05:00:00.000Z,
          "dateSort": 20191102,
          "dateString": "11/02/2019",
          "note": "Hooray everything went to plan",
          "row": 3,
          "status": "G",
          "statusDescription": "Good deployment",
        },
        Object {
          "date": 2019-11-10T06:00:00.000Z,
          "dateSort": 20191110,
          "dateString": "11/10/2019",
          "note": "People are trashing the database",
          "row": 5,
          "status": "O",
          "statusDescription": "Outage",
        },
        Object {
          "date": 2019-12-01T06:00:00.000Z,
          "dateSort": 20191201,
          "dateString": "12/01/2019",
          "note": "This is a note with

Line breaks",
          "row": 6,
          "status": "G",
          "statusDescription": "Good deployment",
        },
        Object {
          "date": 2019-12-02T06:00:00.000Z,
          "dateSort": 20191202,
          "dateString": "12/02/2019",
          "note": "",
          "row": 7,
          "status": "C",
          "statusDescription": "Deployment with Customer Impact",
        },
      ]
    }
    appTitle="Picker"
    endDate={2019-12-15T06:00:00.000Z}
    startDate={2019-11-01T05:00:00.000Z}
  />
  <BuildRow
    appStatus={
      Array [
        Object {
          "date": 2019-11-01T05:00:00.000Z,
          "dateSort": 20191101,
          "dateString": "11/01/2019",
          "note": "Ooopsie!",
          "row": 2,
          "status": "B",
          "statusDescription": "Bad Deployment",
        },
        Object {
          "date": 2019-11-10T06:00:00.000Z,
          "dateSort": 20191110,
          "dateString": "11/10/2019",
          "note": "Some stuff happened but it was ALL GOOD",
          "row": 4,
          "status": "G",
          "statusDescription": "Good deployment",
        },
      ]
    }
    appTitle="Training Hub"
    endDate={2019-12-15T06:00:00.000Z}
    startDate={2019-11-01T05:00:00.000Z}
  />
</Container>
`;
